### PR TITLE
Small improvements to preference comparisons

### PR DESCRIPTION
**Pull request by ejnnr**
These are some changes I made for the reward preprocessing project but hadn't merged so far:
- One small bug fix
- Allow sampling some comparisons at the beginning for pertaining the reward model before training the agent
- Slightly different default configs
- Allow using random actions for some fraction of the sampled trajectories

The last point adds a bit of complexity to the code, so we could also remove that until we actually need it (I used it during debugging and it seems plausible that it could sometimes be helpful, but I don't actually have an example environment where it's needed).

### CODE REVIEW

File: src/imitation/scripts/config/train_preference_comparisons.py [hunk 1]
```diff
@@ -19,11 +19,15 @@
...
     fragment_length = 100  # timesteps per fragment used for comparisons
     total_timesteps = int(1e6)  # total number of environment timesteps
-    total_comparisons = 1000  # total number of comparisons to elicit
+    total_comparisons = 5000  # total number of comparisons to elicit
     # comparisons to gather before switching back to agent training
-    comparisons_per_iteration = 50
+    comparisons_per_iteration = 100
     # factor by which to oversample transitions before creating fragments
-    transition_oversampling = 10
+    transition_oversampling = 1
```
[the remark this answers is not in the document]
Comment (AdamGleave) [reply src/imitation/scripts/config/train_preference_comparisons.py:26]: OK with removing the warning

File: src/imitation/algorithms/preference_comparisons.py [hunk 2]
```diff
@@ -110,6 +110,7 @@ def __init__(
         self,
         algorithm: base_class.BaseAlgorithm,
         reward_fn: Union[rewards_common.RewardFn, reward_nets.RewardNet],
+        random_frac: float = 0.0,
```
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:113]: This seems like a useful feature, but does it belong in the algorithm? One alternative way of implementing this is to have a wrapper for `BaseAlgorithm` that randomizes between a random agent and the agent being trained. WDYT?

File: src/imitation/algorithms/preference_comparisons.py [hunk 3]
```diff
@@ -760,6 +792,10 @@ def __init__(
                 creating fragments. Since fragments are sampled with replacement,
                 this is usually chosen > 1 to avoid having the same transition
                 in too many fragments.
+            initial_comparison_frac: fraction of the total_comparisons argument
+                to train() that will be sampled before the rest of training begins
+                (using the randomly initialized agent). This can be used to pretrain
+                the reward model before the agent is trained on the learned reward.
```
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:798]:
```suggestion
                the reward model before the agent is trained on the learned reward,
                to help avoid irreversibly learning a bad policy from an untrained reward.
```
Paraphrasing the justification given in appendix A of https://arxiv.org/pdf/1706.03741.pdf

File: src/imitation/algorithms/preference_comparisons.py [hunk 4]
```diff
@@ -849,7 +888,7 @@ def train(self, total_timesteps: int, total_comparisons: int) -> Mapping[str, An
             # the desired total number of comparisons, we collect the remainder
             # right at the beginning to pretrain the reward model slightly
             if i == 0:
-                num_pairs += extra_comparisons
+                num_pairs += extra_comparisons + initial_comparisons
```
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:891]: Is this enough? Gathers a bunch of data, but only does a single reward model update AFAICT. Might need an `initial_updates` or make that proportional to the standard number of `self.comparisons_per_iteration`.

**Review by AdamGleave — COMMENTED** [review]
Thanks for the PR and sorry for the slow review! Looks like some solid improvements. I have some questions about whether this is best way of implementing `random_frac` (although seems like a useful feature), and if `initial_frac` is doing enough.

Happy to discuss on this PR or in our next meeting.

File: src/imitation/algorithms/preference_comparisons.py (hunk 4 above)
Comment (ejnnr) [reply src/imitation/algorithms/preference_comparisons.py:896]: Yeah, agreed. I've just made the number of reward training epochs proportional to the number of new comparisons now, wanted to avoid yet another hyperparameter. If we decide we need something like `initial_updates` after all, then maybe we should just implement a schedule that regulates the amount of reward vs policy training (IIRC, the DRLHP paper decays the amount of reward training over time).

My implementation with `epoch_multiplier` seems a bit complicated, maybe we should just pass `epochs` to `reward_trainer.train()`, rather than making it a constructor argument. OTOH, the entire point of having abstract base classes was that the main preference comparison trainer shouldn't make many assumptions about the individual components, and an epoch multiplier is a weaker assumption than an absolute number of epochs.

We could consider just getting rid of the ABC for the reward trainer, or at least making a few more assumptions (such as having epochs over a fixed training set). I think that would simplify the code, but if we'll predictably want to have very different types of reward trainers at some point, I would leave things as they are.

File: src/imitation/scripts/config/train_preference_comparisons.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ejnnr) [reply src/imitation/scripts/config/train_preference_comparisons.py:26]: I noticed I'd already built in the ability to suppress the warning with a parameter to `RandomFragmenter`, so I went with that instead of removing the code entirely. Assuming we implement human feedback at some point, we can then have it turned on by default when using that

File: src/imitation/algorithms/preference_comparisons.py (hunk 2 above)
Comment (ejnnr) [reply src/imitation/algorithms/preference_comparisons.py:113]: That sounds like a good idea because it would make it easy to switch policies within an episode like we discussed. I think we could have two parameters: probability of sticking with the current policy vs randomly picking a new one at each step, and probability of picking each policy (expert or random). That covers purely random trajectories as a special case, so we can replace the current implementation with that.

Unfortunately, I don't think this is really going to simplify the code in `AgentTrainer`. In general, we still need to collect additional expert trajectories, and then collect trajectories with some randomness. Of course we could just take the expert trajectories we already collected during training and fill those up with partially random trajectories, but then the user can't put an upper bound on the proportion of partially random trajectories (we might have collected very few trajectories during training). If you see a way around collecting additional trajectories twice, LMK

Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:113]: >  I think we could have two parameters: probability of sticking with the current policy vs randomly picking a new one at each step, and probability of picking each policy (expert or random). That covers purely random trajectories as a special case, so we can replace the current implementation with that.

Seems easy to implement and strictly more general, so I'm in favour.

> Unfortunately, I don't think this is really going to simplify the code in AgentTrainer

You're right. I was thinking for some reason that we were doing the sampling step twice: once from the RL policy, once random. But of course the RL policy is "sampled" as a side-effect of the RL training process (so is effectively free). I think will need to keep current implementation then. I might rename it from `random_frac` to something like `extra_frac` or `exploration_frac` if the trajectories collected by this are not totally random.

File: src/imitation/algorithms/preference_comparisons.py (hunk 4 above)
Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:897]: I think I misunderstood what the code did previously: I'd assumed we just treated the PreferenceDataset as one big batch and did a single update, but I see now that we do batch it (which is reasonable). I'd have actually thought a priori that a fixed number of epochs for the initial comparison should be fine. Maybe it should be a different number to for non-initial comparisons (on the basis that it's a different size dataset), though.

It's an interesting question how the # of epochs should scale with dataset size. Having it be proportional to size of dataset seems not quite right: it'd mean the number of training updates is quadratic in the number of initial comparisons, as we train for more batches and more epochs the more data we have. I actually usually see the opposite -- when training on very big datasets like with unsupervised natural language models, people will often train for 1 epoch, or less (i.e. only train on random subset of data).

Was adding `epoch_multilpier` necessary to get good performance? If not, inclined to scrap it (sorry for the miscommunication). If so, then I think we should set it to some fixed value for initial vs remaining training (a new hyperparameter, which is annoying, but is OK if it has a sensible default), or implement a schedule as you suggest.

Comment (ejnnr) [reply src/imitation/algorithms/preference_comparisons.py:913]: Since it's been a while, I'm unfortunately not sure whether `epoch_multiplier` was necessary. For now, I've reverted to the original version (i.e. same number of epochs no matter the dataset size). However, I think it makes sense to implement a schedule since this is what the DRLHP paper uses (and we should probably see if replicating that helps with our issues). So I've left the `epoch_multiplier` argument in place in the reward trainer, I'm just no longer using it for now. If you'd rather I remove it for now and re-introduce it when implementing a schedule, I can do that too

Comment (AdamGleave) [reply src/imitation/algorithms/preference_comparisons.py:913]: How are you planning to use `epoch_multiplier` in the future? I see the DRLHP paper uses a schedule for the rate of labeling described in appendix A:

> In the Atari domain we also pretrain the reward predictor for 200 epochs before beginning RL training, to reduce the likelihood of irreversibly learning a bad policy based on an untrained predictor. For the rest of training, labels are fed in at a rate decaying inversely with the number of timesteps; after twice as many timesteps have elapsed, we answer about
half as many queries per unit time.

But this seems to be about the labels added to the dataset, not the number of training epochs, although I could imagine them having a similar effect.

They also use a replay buffer which I think we don't use?

> We maintain a buffer of only the last 3,000 labels and loop over this buffer continuously; this is to ensure that the predictor gives enough weight to new labels (which can represent a shift in distribution) when the total number of labels becomes large

File: src/imitation/algorithms/preference_comparisons.py [hunk 5]
```diff
@@ -394,12 +394,22 @@ def __call__(
...
-    def __init__(self, custom_logger: Optional[imit_logger.HierarchicalLogger] = None):
+    def __init__(
+        self,
+        seed: Optional[int] = None,
+        custom_logger: Optional[imit_logger.HierarchicalLogger] = None,
+    ):
         """Initializes the preference gatherer.
 
         Args:
+            seed: seed for the internal RNG, if applicable
```
Comment (AdamGleave) [inline src/imitation/algorithms/preference_comparisons.py:405]: nice addition, always good to make things easy to replicate

**Review by AdamGleave — APPROVED** [review]
LGTM. I don't understand why we'll need `epoch_multiplier` but if you think it'll be useful after reading my comment then feel free to keep it -- easy enough to remove later anyway.

**Comment by AdamGleave** [comment]
Might be worth adding some code coverage in tests for `random_steps > 0` branch which is currently being missed: https://app.codecov.io/gh/HumanCompatibleAI/imitation/compare/371/tree/src/imitation/algorithms/preference_comparisons.py

The other reductions in coverage seem innocuous from a quick skim, so if adding that test doesn't resolve the coverage requirement I'm happy to override that check.

File: src/imitation/algorithms/preference_comparisons.py (hunk 4 above)
Comment (ejnnr) [reply src/imitation/algorithms/preference_comparisons.py:913]: You're right, I misremembered what they use a schedule for. I don't have a good intuition for which aspects will be important to make preference comparisons works, so it might be best to stay very close to the DRLHP paper for now. That means I've added a new parameter that specifies the `epoch_multiplier` for pretaining (i.e. we can use that to train for 200 epochs in the first round but then train less during all remaining iterations). Can easily generalize this to a schedule if we ever need to.

I'll limit the size of our dataset and introduce the preference collection schedule in another PR.

File: src/imitation/algorithms/preference_comparisons.py (hunk 2 above)
Comment (ejnnr) [reply src/imitation/algorithms/preference_comparisons.py]: Implemented the more general version now and renamed to `exploration_frac`

**Comment by ejnnr** [comment]
I think I've finished everything now, but requested re-review since there are quite a few changes (mainly the new `ExplorationWrapper` to implement the more general partially random policies we decided on).

Re `epoch_multiplier`: you're right, the DRLHP paper doesn't have a scheduler for epochs. I still left it in because I'm now using it to implement a longer first round of training (e.g. their 200 epochs). No plans to implement a scheduler for epochs for now, though it would be easy if we ever want one

File: src/imitation/policies/exploration_wrapper.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+from typing import Optional
+
+import numpy as np
+from stable_baselines3.common import vec_env
+
+from imitation.data import rollout
+
+
+class ExplorationWrapper:
+    """Wraps a PolicyCallable to create a partially randomized version.
```
Comment (AdamGleave) [inline src/imitation/policies/exploration_wrapper.py:12]: I think we could make this a more general `MixtureWrapper` without much implementation complexity: take a sequence of `policies` and `probs` (defining a categorical probability distribution), and `stay_prob`.

`ExplorationWrapper` is then just special-case where `policies = [policy, self._random_policy]` and `probs = [random_prob, 1 - random_prob]`.

I don't have a concrete use case in mind for this, so think it's fine to leave it as-is, but feels like a feature that's nice to have around. E.g. what if we want a random policy with some different distribution over actions?

File: src/imitation/policies/exploration_wrapper.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+    This wrapper randomly switches between two policies: the wrapped policy,
+    and a uniformly random one. After each action, the current policy is kept
```
Comment (AdamGleave) [inline src/imitation/policies/exploration_wrapper.py:15]: Not necessarily uniform, whatever the action spaces `sample()` method is (e.g. for continuous unbounded distributions it's Gaussian).

File: src/imitation/policies/exploration_wrapper.py [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        return np.stack(acts, axis=0)
+
+    def _switch(self):
+        """Pick a new policy at random."""
+        if self.rng.rand() < self.random_prob:
+            self.current_policy = self._random_policy
+        else:
+            self.current_policy = self.wrapped_policy
+
+    def __call__(self, obs: np.ndarray):
```
Comment (AdamGleave) [inline src/imitation/policies/exploration_wrapper.py:60]:
```suggestion
    def __call__(self, obs: np.ndarray) -> np.ndarray:
```

File: src/imitation/policies/exploration_wrapper.py [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+        self.current_policy = policy
+        # Choose the initial policy at random
+        self._switch()
+
+    def _random_policy(self, states):
+        acts = [self.venv.action_space.sample() for _ in range(len(states))]
+        return np.stack(acts, axis=0)
+
+    def _switch(self):
```
Comment (AdamGleave) [inline src/imitation/policies/exploration_wrapper.py:53]:
```suggestion
    def _switch(self) -> None:
```

File: src/imitation/policies/exploration_wrapper.py [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        self.venv = venv
+
+        self.rng = np.random.RandomState(seed)
+        self.venv.action_space.seed(seed)
+
+        self.current_policy = policy
+        # Choose the initial policy at random
+        self._switch()
+
+    def _random_policy(self, states):
```
Comment (AdamGleave) [inline src/imitation/policies/exploration_wrapper.py:49]:
```suggestion
    def _random_policy(self, states: np.ndarray) -> np.ndarray:
```
Is it states or obs?

File: tests/algorithms/test_preference_comparisons.py [hunk 11]
```diff
@@ -245,3 +245,21 @@ def test_store_and_load_preference_dataset(agent_trainer, fragmenter, tmp_path):
 
         assert preference == loaded_preference
         _check_trajs_equal(fragments, loaded_fragments)
+
```
Comment (AdamGleave) [inline tests/algorithms/test_preference_comparisons.py:248]: Weak preference but I'd consider adding one or two more tests of `ExplorationWrapper`, mainly because it actually seems amenable to unit testing. Ideas:
  1. If `random_prob == 0`, is `self.current_policy` always `wrapped_policy`, and if `== 1`, always `random_policy`? If `0.5`, is it (roughly) equal after ~100 calls to switch? Can make this deterministic by setting seed.
  2. Check `__call__` always returns things inside `venv.action_space`

**Review by AdamGleave — COMMENTED** [review]
Thanks for the changes! This looks basically ready. I have some suggestions for changes, but I don't think you need to incorporate all of them (e.g. generalizing `ExplorationWrapper`) if you want to just move on.

File: src/imitation/algorithms/preference_comparisons.py [hunk 12]
```diff
@@ -906,11 +932,16 @@ def train(
             # Gather new preferences #
             ##########################
             num_pairs = self.comparisons_per_iteration
-            # if the number of comparisons per iterations doesn't exactly divide
+            epoch_multiplier = 1.0
+            # If the number of comparisons per iterations doesn't exactly divide
             # the desired total number of comparisons, we collect the remainder
-            # right at the beginning to pretrain the reward model slightly
+            # right at the beginning to pretrain the reward model slightly.
```
[the remark this answers is not in the document]
Comment (ejnnr) [reply src/imitation/algorithms/preference_comparisons.py:938]: Good point, I've added a comment and a note in the doctoring for `comparisons_per_iteration`. With the hyperparameters I've been using so far, this shouldn't be a huge problem (e.g. for 5000 total comparisons, 100 comparisons per iteration, this is at most 2% of the total comparisons, whereas `initial_comparison_frac` is typically more like 10% or 20%). But could be an issue if `initial_comparison_frac` is 0.

If I implement a scheduler, I'll have to rewrite this logic anyway, I'll see then whether it's easy to do it correctly right away.

File: src/imitation/policies/exploration_wrapper.py (hunk 6 above)
Comment (ejnnr) [reply src/imitation/policies/exploration_wrapper.py:12]: I had the same thought but I think this should be easy to change later if we ever need it, so I'll stay with the simple  special case for now

File: src/imitation/policies/exploration_wrapper.py (hunk 10 above)
Comment (ejnnr) [reply src/imitation/policies/exploration_wrapper.py:49]: I went with `obs` now, I think `imitation` uses that (almost?) everywhere else

File: tests/algorithms/test_preference_comparisons.py (hunk 11 above)
Comment (ejnnr) [reply tests/algorithms/test_preference_comparisons.py:248]: Done
